File: src/ai/rag/rag_pipeline.py
import json
import time
import requests
from typing import List, Dict, Any, Optional
from src.ai.config import config
from src.ai.rag.vector_store import PythonVectorStore
from src.ai.schemas.alumni_schemas import AlumniSearchResult, AlumniMatch, ExtractedIntent

class PythonRAGPipeline:
    def __init__(self):
        self.vector_store = PythonVectorStore()

    def execute_rag_search(
        self, 
        user_query: str, 
        student_profile_skills: List[str],
        db: List[Dict[str, Any]], 
        top_k: int = 5
    ) -> AlumniSearchResult:
        start_time = time.time()
        
        if not self.vector_store.is_indexed or len(self.vector_store.documents) != len(db):
            self.vector_store.index_documents(db)

        combined_query = f"{user_query} {' '.join(student_profile_skills)}"
        retrieved_results = self.vector_store.search(combined_query, top_k=top_k)

        if not retrieved_results:
            return AlumniSearchResult(
                query=user_query,
                extractedIntent=ExtractedIntent(studentSkills=student_profile_skills, targetDomain="General", targetCompanies=[]),
                matches=[]
            )

        top_candidates = [r["document"] for r in retrieved_results]

        # Call Gemini or Ollama if configured
        if config.GEMINI_API_KEY:
            llm_res = self._call_gemini_rag(user_query, student_profile_skills, top_candidates)
            if llm_res and len(llm_res.matches) > 0:
                return llm_res

        # High-Speed Vector Fallback
        matches = []
        for r in retrieved_results:
            doc = r["document"]
            matched_skills = r["matchedTerms"] if r["matchedTerms"] else doc.get("skills", ["Engineering"])[:3]
            matches.append(
                AlumniMatch(
                    alumniId=str(doc.get("id", "")),
                    name=doc.get("name", "Alumnus"),
                    role=doc.get("role") or doc.get("title") or "Engineer",
                    company=doc.get("company", ""),
                    matchedSkills=matched_skills,
                    matchScore=r["matchScore"],
                    matchPercentage=f"{r['matchScore']}%",
                    isSimilarRecommendation=False,
                    reason=f"Vector RAG matched similarity ({int(r['similarity']*100)}%) with background as {doc.get('role')} at {doc.get('company')}."
                )
            )

        return AlumniSearchResult(
            query=user_query,
            extractedIntent=ExtractedIntent(studentSkills=student_profile_skills, targetDomain="Software Engineering", targetCompanies=[]),
            matches=matches
        )

    def _call_gemini_rag(
        self, 
        user_query: str, 
        skills: List[str], 
        top_candidates: List[Dict[str, Any]]
    ) -> Optional[AlumniSearchResult]:
        prompt = f"""
You are the RAG Alumni Matcher for ConnectEd.
Student Query / Skills: "{user_query}" {json.dumps(skills)}

Retrieved Candidates ({len(top_candidates)} items):
{json.dumps([{
  "id": c.get("id"),
  "name": c.get("name"),
  "role": c.get("role") or c.get("title"),
  "company": c.get("company"),
  "skills": c.get("skills"),
  "bio": c.get("bio")
} for c in top_candidates], indent=2)}

INSTRUCTIONS:
Select top matching alumni and provide clear explanations.
Return ONLY valid JSON matching this schema:
{{
  "query": "{user_query}",
  "extractedIntent": {{
    "studentSkills": {json.dumps(skills)},
    "targetDomain": "Domain Name",
    "targetCompanies": []
  }},
  "matches": [
    {{
      "alumniId": "alum-id",
      "name": "Name",
      "role": "Role",
      "company": "Company",
      "matchedSkills": ["skill"],
      "matchScore": 92,
      "matchPercentage": "92%",
      "isSimilarRecommendation": false,
      "reason": "Human explanation"
    }}
  ]
}}
"""
        payload = {
            "contents": [{"parts": [{"text": prompt}]}],
            "generationConfig": {"responseMimeType": "application/json", "temperature": 0.1}
        }

        for model_name in config.GEMINI_MODELS:
            try:
                url = f"https://generativelanguage.googleapis.com/v1beta/models/{model_name}:generateContent?key={config.GEMINI_API_KEY}"
                res = requests.post(url, json=payload, timeout=20)
                if res.status_code == 200:
                    data = res.json()
                    text = data["candidates"][0]["content"]["parts"][0]["text"]
                    parsed = json.loads(text)
                    return AlumniSearchResult(**parsed)
            except Exception:
                pass
        return None

global_rag_pipeline = PythonRAGPipeline()

File: src/ai/rag/ragPipeline.js
/**
 * ConnectEd RAG Pipeline Executor
 * 
 * Performs 3-step Retrieval-Augmented Generation:
 * 1. RETRIEVE: Vector Store K-NN retrieval for Top-K relevant candidates (< 5ms)
 * 2. AUGMENT: Constructs concise RAG context prompt with Top-K items only
 * 3. GENERATE: Invokes LLM (Gemini / Ollama / Local Fallback) for explainable matches
 */

import { VectorStore } from './vectorStore.js';
import { generateStructuredJson } from '../utils/llmClient.js';
import { AlumniSearchResultSchema } from '../schemas/alumniSearchSchema.js';

export class RAGPipeline {
  constructor(vectorStore = null) {
    this.vectorStore = vectorStore || new VectorStore();
  }

  /**
   * Initializes or updates the vector index with the latest alumni database.
   */
  initializeIndex(alumniList) {
    if (Array.isArray(alumniList) && alumniList.length > 0) {
      this.vectorStore.indexDocuments(alumniList);
    }
  }

  /**
   * Executes RAG semantic alumni search.
   * 
   * @param {string} userQuery - Student's natural language query or profile description
   * @param {Array} alumniDb - Complete alumni list from PostgreSQL or mock data
   * @param {number} topK - Number of candidates to retrieve for LLM augmentation (default: 5)
   */
  async executeSearch(userQuery, alumniDb = [], topK = 5) {
    const startTime = Date.now();

    // Ensure Vector Index is populated
    if (!this.vectorStore.isIndexed || this.vectorStore.documents.length === 0) {
      this.initializeIndex(alumniDb);
    }

    // Step 1: RETRIEVE Top-K Vector Candidates via Cosine Similarity
    const retrievedResults = this.vectorStore.search(userQuery, topK);
    const retrievalTimeMs = Date.now() - startTime;

    if (retrievedResults.length === 0) {
      return {
        query: userQuery,
        extractedIntent: { studentSkills: [], targetDomain: "General" },
        matches: [],
        meta: { retrievalTimeMs, totalCandidatesEvaluated: alumniDb.length }
      };
    }

    const topKCandidates = retrievedResults.map(r => r.document);

    // Step 2: AUGMENT Context Prompt with ONLY Top-K Candidates (90% Token Reduction)
    const ragPrompt = `
You are the AI Semantic Alumni RAG Matcher for ConnectEd.
Student Query / Goal: "${userQuery}".

Retrieved Top-${retrievedResults.length} Candidate Alumni (filtered via Vector Cosine Similarity Search):
${JSON.stringify(topKCandidates.map(c => ({
  id: c.id,
  name: c.name,
  role: c.role || c.title,
  company: c.company,
  skills: c.skills,
  bio: c.bio,
  availability: c.availability
})), null, 2)}

INSTRUCTIONS:
1. Analyze the student query and candidate profiles.
2. Select the best matching alumni from the retrieved list.
3. For each selected candidate, provide a clear, human-understandable reason explaining WHY they match.
4. Calculate realistic match scores (60-98).

Return ONLY valid JSON matching this schema:
{
  "query": "${userQuery}",
  "extractedIntent": {
    "studentSkills": ["skill1", "skill2"],
    "targetDomain": "Domain Name"
  },
  "matches": [
    {
      "alumniId": "alum-id",
      "name": "Name",
      "role": "Role",
      "company": "Company",
      "matchedSkills": ["skill"],
      "matchScore": 92,
      "matchPercentage": "92%",
      "reason": "Human explanation"
    }
  ]
}
`;

    // Step 3: GENERATE Explainable Match Reasonings via LLM
    try {
      const llmResult = await generateStructuredJson(ragPrompt, AlumniSearchResultSchema);
      if (llmResult && llmResult.matches && llmResult.matches.length > 0) {
        return {
          ...llmResult,
          meta: {
            retrievalTimeMs,
            totalCandidatesEvaluated: alumniDb.length,
            retrievedCandidateCount: topKCandidates.length,
            ragEnabled: true
          }
        };
      }
    } catch (err) {
      console.warn("RAG LLM generation fallback triggered:", err.message);
    }

    // Step 4: Fallback RAG Engine (High-speed vector score formatting when offline)
    const fallbackMatches = retrievedResults.map(r => {
      const doc = r.document;
      return {
        alumniId: String(doc.id),
        name: doc.name,
        role: doc.role || doc.title,
        company: doc.company,
        matchedSkills: r.matchedTerms.length > 0 ? r.matchedTerms : (doc.skills ? doc.skills.slice(0, 3) : ['Engineering']),
        matchScore: r.matchScore,
        matchPercentage: `${r.matchScore}%`,
        reason: doc.matchReason || `Expertise in ${(r.matchedTerms.length > 0 ? r.matchedTerms : (doc.skills || [])).slice(0, 3).join(', ')} as ${doc.role || doc.title} at ${doc.company} aligns with your query.`
      };
    });

    return {
      query: userQuery,
      extractedIntent: {
        studentSkills: retrievedResults[0]?.matchedTerms || ["Engineering"],
        targetDomain: topKCandidates[0]?.role || "Software Engineering"
      },
      matches: fallbackMatches,
      meta: {
        retrievalTimeMs,
        totalCandidatesEvaluated: alumniDb.length,
        retrievedCandidateCount: topKCandidates.length,
        ragEnabled: true,
        fallbackMode: true
      }
    };
  }
}

export const globalRAGPipeline = new RAGPipeline();
